ynr/apps/bulk_adding/views/parties.py

```diff
@@ -0,0 +1,217 @@
+from django.http import HttpResponseRedirect
+from django.views.generic import FormView, TemplateView
+from popolo.models import Identifier
+
+from bulk_adding import forms, helpers
+from candidates.models import MembershipExtra, OrganizationExtra, PersonExtra
+from elections.models import Election
+
+
+class BasePartyBulkAddView(LoginRequiredMixin, TemplateView):
+    def get_election(self):
+        if not hasattr(self, '_election_obj'):
+            self._election_obj = Election.objects.get(
+                slug=self.kwargs['election'])
+        return self._election_obj
+
+    def get_party(self):
+        identifier = Identifier.objects.get(
+            scheme='electoral-commission',
+            identifier=self.kwargs['party_id']
+        )
+        return identifier.content_object
+
+    def get_pee_qs(self, election):
+        qs = election.postextraelection_set.all()
+        qs = qs.order_by('postextra__base__label')
+        return qs
+
+
+class SelectPartyForm(BasePartyBulkAddView, FormView):
+    template_name = "bulk_add/parties/select_party.html"
+    form_class = forms.SelectPartyForm
+
+    def get_context_data(self, **kwargs):
+        context = super(SelectPartyForm, self).get_context_data(**kwargs)
+        context['election_obj'] = self.get_election()
+        return context
+
+    def get_form_kwargs(self):
+        kwargs = super(SelectPartyForm, self).get_form_kwargs()
+        kwargs['election'] = self.get_election()
+        return kwargs
+
+    def form_invalid(self, form):
+        form_data = dict(form.data)
+        for k, v in form_data.items():
+            if not k == "csrfmiddlewaretoken":
+                del form_data[k]
+            form.data = form_data
+        return self.render_to_response(self.get_context_data(form=form))
+
+    def form_valid(self, form):
+        org = OrganizationExtra.objects.get(
+            base__pk=form.cleaned_data['party'])
+        org_ec_id = org.base.identifiers.get(scheme='electoral-commission')
+        return HttpResponseRedirect(
+            reverse('bulk_add_by_party', kwargs={
+                'election': self.get_election().slug,
+                'party_id': org_ec_id.identifier,
+            })
+        )
+
+
+class BulkAddPartyView(BasePartyBulkAddView):
+    template_name = "bulk_add/parties/add_form.html"
+
+    def get_context_data(self, **kwargs):
+        context = super(BulkAddPartyView, self).get_context_data(**kwargs)
+        context['election_obj'] = self.get_election()
+        context['party'] = self.get_party()
+        context['form'] = forms.AddByPartyForm()
+        posts = []
+        qs = self.get_pee_qs(context['election_obj'])
+        for pee in qs:
+            factory = django_forms.formset_factory(
+                forms.NameOnlyPersonForm,
+                # Assume 3 winners if we have no other info.
+                # If someone reports this, we can update EE and re-import
+                extra=pee.winner_count or 3,
```

Comment:
I _think_ this  is a sensible place to leave a line comment for this:

Where we are deciding how many slots to show, should this be `seats_contested` or `seats_total` (depending on how reliable we think either of these numbers are?) rather than `winner_count`? Also is 3 definitely a good fallback value if we don't know? There are definitely local elections where a single party will field 4 or 5 candidates in a single ward. For example if you look at https://democracy.swansea.gov.uk/mgMemberIndex.aspx?FN=WARD&VW=LIST&PIC=0&LLL=0 there are wards in Swansea which are represented by 4 or 5 councillors and we had all-up elections last year. The 3 major parties definitely all had 4 candidates on the ballot in my ward.

Reply:
We don't store `seats_total` in a meaningful way in YNR (this is a possible future feature. The closest we get is `people_elected_per_post` on `Election`, but that's clearly not useful at all – I guess it works for national elections where the default isn't 1, but that's not something we'd want to rely on.

At import time `seats_contested` is converted to `winner_count`, or at least it will be [when this commit is merged](https://github.com/DemocracyClub/yournextrepresentative/pull/445/commits/f80a54de8f310e72a91845ac96807b6535d37729).

As for the default, yeah, I can make it 4 or 5 no problem.  I wanted to keep the page size smallish, but I think you're right to point out that 3 is too low.

Reply:
> At import time `seats_contested` is converted to `winner_count`

Sorted then :+1: That is a sensible number to use :)